File: front-end/src/app/domain/models/game/Player.ts
import {User} from "../user.model";
import {JailBreakCard} from "./card/impl/JailBreakCard";
import {OwnableLocation} from "./board/location/OwnableLocation";
import {CardUtil} from "./card/CardUtil";
import {LocationUtil} from "./board/location/LocationUtil";
import {OwnableLocationBid} from "./board/location/OwnableLocationBid";
import {JailBreakCardBid} from "./card/impl/JailBreakCardBid";
import {GameAssetBidUtil} from "./GameAssetBidUtil";

export class Player {
  private _id: number;
  private _user: User;
  private _money: number;
  private _position: number;

  /**
   * 0: Meaning that the person is not in jail.
   * 1: First round in Jail.
   * 2: Second round in Jail.
   * ...
   */
  private _roundsInJail: number;
  private _properties: OwnableLocation[];
  private _cards: JailBreakCard[];
  private _gameAdmin: boolean;
  private _bankrupt: boolean;
  private _turnNumber: number;

  public constructor(
    id: number,
    user: User,
    money: number,
    position: number,
    roundsInJail: number,
    properties: OwnableLocation[],
    cards: JailBreakCard[],
    gameAdmin: boolean,
    bankrupt: boolean,
    turnNumber: number,
    receivedLocationBids: OwnableLocationBid[],
    receivedJailBreakCardBids: JailBreakCardBid[],
    performedLocationBids: OwnableLocationBid[],
    performedJailBreakCardBids: JailBreakCardBid[]
  ) {
    this._id = id;
    this._user = user;
    this._money = money;
    this._position = position;
    this._roundsInJail = roundsInJail;
    this._properties = properties;
    this._cards = cards;
    this._gameAdmin = gameAdmin;
    this._bankrupt = bankrupt;
    this._turnNumber = turnNumber;
    this._receivedLocationBids = receivedLocationBids;
    this._receivedJailBreakCardBids = receivedJailBreakCardBids;
    this._performedLocationBids = performedLocationBids;
    this._performedJailBreakCardBids = performedJailBreakCardBids;
  }

  private _receivedLocationBids: OwnableLocationBid[];

  get receivedLocationBids(): OwnableLocationBid[] {
    return this._receivedLocationBids;
  }

  set receivedLocationBids(value: OwnableLocationBid[]) {
    this._receivedLocationBids = value;
  }

  private _receivedJailBreakCardBids: JailBreakCardBid[];

  get receivedJailBreakCardBids(): JailBreakCardBid[] {
    return this._receivedJailBreakCardBids;
  }

  set receivedJailBreakCardBids(value: JailBreakCardBid[]) {
    this._receivedJailBreakCardBids = value;
  }

  private _performedLocationBids: OwnableLocationBid[];

  get performedLocationBids(): OwnableLocationBid[] {
    return this._performedLocationBids;
  }

  set performedLocationBids(value: OwnableLocationBid[]) {
    this._performedLocationBids = value;
  }

  private _performedJailBreakCardBids: JailBreakCardBid[];

  get performedJailBreakCardBids(): JailBreakCardBid[] {
    return this._performedJailBreakCardBids;
  }

  set performedJailBreakCardBids(value: JailBreakCardBid[]) {
    this._performedJailBreakCardBids = value;
  }

  public get id(): number {
    return this._id;
  }

  public get user(): User {
    return this._user;
  }

  public get money(): number {
    return this._money;
  }

  public set money(value: number) {
    this._money = value;
  }

  public get position(): number {
    return this._position;
  }

  public set position(value: number) {
    this._position = value;
  }

  public get roundsInJail(): number {
    return this._roundsInJail;
  }

  public set roundsInJail(value: number) {
    this._roundsInJail = value;
  }

  public get properties(): OwnableLocation[] {
    return this._properties;
  }

  public set properties(value: OwnableLocation[]) {
    this._properties = value;
  }

  public get cards(): JailBreakCard[] {
    return this._cards;
  }

  public set cards(value: JailBreakCard[]) {
    this._cards = value;
  }

  public get gameAdmin(): boolean {
    return this._gameAdmin;
  }

  public get bankrupt(): boolean {
    return this._bankrupt;
  }

  public set bankrupt(value: boolean) {
    this._bankrupt = value;
  }

  public get turnNumber(): number {
    return this._turnNumber;
  }

  public static trueCopy(player: Player): Player {
    return new Player(
      player.id,
      User.trueCopy(player.user),
      player.money,
      player.position,
      player.roundsInJail,
      player.properties.map(loc => <OwnableLocation>LocationUtil.trueCopy(loc)),
      player.cards.map(item => <JailBreakCard>CardUtil.trueCopy(item)),
      player.gameAdmin,
      player.bankrupt,
      player.turnNumber,
      <OwnableLocationBid[]>player.receivedLocationBids.map(GameAssetBidUtil.trueCopy),
      <JailBreakCardBid[]>player.receivedJailBreakCardBids.map(GameAssetBidUtil.trueCopy),
      <OwnableLocationBid[]>player.performedLocationBids.map(GameAssetBidUtil.trueCopy),
      <JailBreakCardBid[]>player.performedJailBreakCardBids.map(GameAssetBidUtil.trueCopy)
    );
  }

  public static compareRank(p1: Player, p2: Player): number {
    let p1NetWorth: number = p1.money;
    p1.properties.forEach(property => p1NetWorth += property.initialPurchasePrice);

    let p2NetWorth: number = p2.money;
    p2.properties.forEach(property => p2NetWorth += property.initialPurchasePrice);

    return p2NetWorth - p1NetWorth;
  }

  public toJson(): object {
    return {
      "id": this.id,
      "user": this.user.toJson(),
      "money": this.money,
      "position": this.position,
      "roundsInJail": this.roundsInJail,
      "properties": this.properties.map(location => location.toJson()),
      "mortgagedProperties": this.properties.map(location => location.toJson()),
      "cards": this.cards.map(item => item.toJson()),
      "gameAdmin": this.gameAdmin,
      "bankrupt": this.bankrupt,
      "turnNumber": this.turnNumber,
      "receivedLocationBids": this._receivedLocationBids.map(item => item.toJson()),
      "receivedJailBreakCardBids": this._receivedJailBreakCardBids.map(item => item.toJson()),
      "performedLocationBids": this._performedLocationBids.map(item => item.toJson()),
      "performedJailBreakCardBids": this._performedJailBreakCardBids.map(item => item.toJson())
    }
  }
}

File: back-end/src/main/java/nl/hva/backend/domain/models/game/Player.java
package nl.hva.backend.domain.models.game;

import com.fasterxml.jackson.annotation.JsonIdentityInfo;
import com.fasterxml.jackson.annotation.ObjectIdGenerators;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import nl.hva.backend.domain.models.game.board.location.Location;
import nl.hva.backend.domain.models.game.board.location.OwnableLocation;
import nl.hva.backend.domain.models.game.card.impl.JailBreakCard;
import nl.hva.backend.domain.models.user.User;

import javax.persistence.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.function.Predicate;

@Entity
@Getter
@Setter
@NoArgsConstructor

@JsonIdentityInfo(
        generator = ObjectIdGenerators.PropertyGenerator.class,
        property = "id",
        scope = Player.class
)
public class Player {
    public final static Predicate<Player> IS_PLAYER_BANKRUPT_PREDICATE = Player::isBankrupt;

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private long id;

    @ManyToOne
    private GameSession game;

    @ManyToOne
    private User user;

    //region state attributes
    private double money;

    /**
     * Board has 40 positions.
     * <p>
     * Pos 0: Placeholder for bankrupt players.
     * Pos 1: Go
     */
    private int position = 1;

    /**
     * 0: Meaning that the person is not in jail.
     * 1: First round in Jail.
     * 2: Second round in Jail.
     * ...
     */
    private int roundsInJail;

    private int turnNumber;

    private boolean gameAdmin;

    private boolean bankrupt;
    //endregion

    //region property attributes
    @ManyToMany
    private List<OwnableLocation> properties;

    @ManyToMany
    private List<JailBreakCard> cards;
    //endregion

    //region bids attributes
    @OneToMany
    private List<OwnableLocation.OwnableLocationBid> receivedLocationBids;
    @OneToMany
    private List<JailBreakCard.JailBreakCardBid> receivedJailBreakCardBids;
    @OneToMany
    private List<OwnableLocation.OwnableLocationBid> performedLocationBids;
    @OneToMany
    private List<JailBreakCard.JailBreakCardBid> performedJailBreakCardBids;
    //endregion

    //region addition to collection helper methods
    public void addReceivedBid(OwnableLocation.OwnableLocationBid bid) {
        if (bid == null) return;

        if (getReceivedLocationBids() == null
                || !(getReceivedLocationBids() instanceof ArrayList)
        ) setReceivedLocationBids(new ArrayList<>(Objects.requireNonNullElse(getReceivedLocationBids(), List.of())));

        getReceivedLocationBids().add(bid);
    }

    public void addReceivedBid(JailBreakCard.JailBreakCardBid bid) {
        if (bid == null) return;

        if (getReceivedJailBreakCardBids() == null
                || !(getReceivedJailBreakCardBids() instanceof ArrayList)
        ) setReceivedJailBreakCardBids(
                new ArrayList<>(Objects.requireNonNullElse(getReceivedJailBreakCardBids(), List.of())));

        getReceivedJailBreakCardBids().add(bid);
    }

    public void addPerformedBid(OwnableLocation.OwnableLocationBid bid) {
        if (bid == null) return;

        if (getPerformedLocationBids() == null
                || !(getPerformedLocationBids() instanceof ArrayList)
        ) setPerformedLocationBids(new ArrayList<>(Objects.requireNonNullElse(getPerformedLocationBids(), List.of())));

        getPerformedLocationBids().add(bid);
    }

    public void addPerformedBid(JailBreakCard.JailBreakCardBid bid) {
        if (bid == null) return;

        if (getPerformedJailBreakCardBids() == null
                || !(getPerformedJailBreakCardBids() instanceof ArrayList)
        ) setPerformedJailBreakCardBids(
                new ArrayList<>(Objects.requireNonNullElse(getPerformedJailBreakCardBids(), List.of())));

        getPerformedJailBreakCardBids().add(bid);
    }

    public void add(OwnableLocation location) {
        if (location == null) return;

        if (getProperties() == null
                || !(getProperties() instanceof ArrayList)
        ) setProperties(new ArrayList<>(Objects.requireNonNullElse(getProperties(), List.of())));

        getProperties().add(location);
    }

    public void add(JailBreakCard card) {
        if (card == null) return;

        if (getCards() == null
                || !(getCards() instanceof ArrayList)
        ) setCards(new ArrayList<>(Objects.requireNonNullElse(getCards(), List.of())));

        getCards().add(card);
    }
    //endregion

    public int compareRankTo(Player p2) {
        double p1NetWorth = this.getMoney()
                + this.getProperties()
                .stream()
                .mapToDouble(OwnableLocation::getInitialPurchasePrice)
                .sum();

        double p2NetWorth = p2.getMoney()
                + p2.getProperties()
                .stream()
                .mapToDouble(OwnableLocation::getInitialPurchasePrice)
                .sum();

        return Double.compare(p1NetWorth, p2NetWorth);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Location location)) return false;
        return getId() == location.getId();
    }

    @Override
    public int hashCode() {
        return Long.hashCode(getId());
    }

}